src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -296,7 +296,14 @@ static SymbolLookup systemLookup() {
     static MemorySegment toCString(String str, SegmentAllocator allocator) {
         Objects.requireNonNull(str);
         Objects.requireNonNull(allocator);
-        return toCString(str.getBytes(), allocator);
+        return toCString(addNullTerminator(str).getBytes(), allocator);
+    }
+
+    private static String addNullTerminator(String str) {
```

Comment:
This is gonna allocate another string in an already allocation-heavy code. Wouldn't it be better to just add the correct termination sequence in the segment? I suggest running StrLen benchmark before/after to make sure string conversion performance isn't negatively impacted.

Reply:
I'll check the benchmark.

I looked at the CharSet API, and AFAICS the only way to get the number of bytes for the null terminator would be to create a String with the value `"\0"` then encode that and check the resulting `byte[]`. I thought going the string concat route would have better chances of being optimized.

Reply:
The benchmark results are as follows:

Before:

```
Benchmark                        (size)  Mode  Cnt    Score    Error  Units
StrLenTest.panama_strlen_prefix       5  avgt   30  124.874 � 15.751  ns/op
StrLenTest.panama_strlen_prefix      20  avgt   30  131.683 �  6.011  ns/op
StrLenTest.panama_strlen_prefix     100  avgt   30  161.046 � 10.580  ns/op
```

After:
```
Benchmark                        (size)  Mode  Cnt    Score   Error  Units
StrLenTest.panama_strlen_prefix       5  avgt   30  130.758 � 5.691  ns/op
StrLenTest.panama_strlen_prefix      20  avgt   30  145.012 � 6.804  ns/op
StrLenTest.panama_strlen_prefix     100  avgt   30  179.992 � 6.457  ns/op
```

I think C2 should be able to eliminate the intermediate string, but there's still a slight regression.

> Wouldn't it be better to just add the correct termination sequence in the segment?

The problem is finding out how many bytes to add. Looking at Charset again, there really doesn't seem to be a way to get the number of bytes per character. The closest seems to be `charset.newEncoder().averageBytesPerChar()`. I'll ask around as well.